src/realm/object-store/sync/app.cpp

```diff
@@ -1048,51 +1050,72 @@ std::string App::function_call_url_path() const
-        log("App: call_function: %1 service_name: %2 args_bson: %3", name, service_name2, query_stg);
+        log("App: call_function: %1 service_name: %2 args_bson: %3", name, service_name2, args_ejson);
     }
-    auto handler = [self = shared_from_this(), name = name, service_name = service_name2,
-                    completion = std::move(completion)](const Response& response) {
-        if (auto error = AppUtils::check_for_errors(response)) {
-            self->log("App: call_function: %1 service_name: %2 -> %3 ERROR: %4", name, service_name,
-                      response.http_status_code, error->message);
-            return completion(util::none, error);
-        }
-        util::Optional<Bson> body_as_bson;
-        try {
-            body_as_bson = bson::parse(response.body);
-            if (self->would_log()) {
-                self->log("App: call_function: %1 service_name: %2 - results: %3", name, service_name,
-                          body_as_bson ? body_as_bson->to_string() : "<none>");
-            }
-        }
-        catch (const std::exception& e) {
-            self->log("App: call_function: %1 service_name: %2 - error parsing result: %3", name, service_name,
-                      e.what());
-            return completion(util::none, AppError(make_error_code(JSONErrorCode::bad_bson_parse), e.what()));
-        };
-        completion(std::move(body_as_bson), util::none);
-    };
-
-    BsonDocument args{{"arguments", args_bson}, {"name", name}};
 
+    auto args = util::format("{\"arguments\":$1,\"name\":$2", args_ejson, nlohmann::json(name).dump());
     if (service_name) {
-        args["service"] = *service_name;
+        args += ",\"service\":";
+        args += nlohmann::json(*service_name).dump();
     }
+    args += '}';
 
     do_authenticated_request(
-        Request{HttpMethod::post, function_call_url_path(), m_request_timeout_ms, {}, Bson(args).toJson(), false},
-        user, std::move(handler));
+        Request{HttpMethod::post, function_call_url_path(), m_request_timeout_ms, {}, std::move(args), false}, user,
+        [self = shared_from_this(), name = name, service_name = std::move(service_name2),
+         completion = std::move(completion)](const Response& response) {
+            if (auto error = AppUtils::check_for_errors(response)) {
+                self->log("App: call_function: %1 service_name: %2 -> %3 ERROR: %4", name, service_name,
+                          response.http_status_code, error->message);
+                return completion({}, error);
+            }
+            completion(response.body, util::none);
+        });
+}
+
+void App::call_function(const std::shared_ptr<SyncUser>& user, const std::string& name, const BsonArray& args_bson,
+                        const Optional<std::string>& service_name,
+                        UniqueFunction<void(Optional<Bson>&&, Optional<AppError>)>&& completion)
+{
+    auto service_name2 = service_name ? *service_name : "<none>";
```

Comment:
I know you copied this from elsewhere in the file, but it'd be great to not name variables `somethingsomething2`. How about naming `service_name` -> `maybe_service_name` and making this `auto service_name = maybe_service_name.value_or("<none>")`